internal/db/db_certificates_test.go

```diff
@@ -0,0 +1,96 @@
+package db_test
+
+import (
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/canonical/notary/internal/db"
+)
+
+func TestCertificatesEndToEnd(t *testing.T) {
+	tempDir := t.TempDir()
+	db, err := db.NewDatabase(filepath.Join(tempDir, "db.sqlite3"))
+	if err != nil {
+		t.Fatalf("Couldn't complete NewDatabase: %s", err)
+	}
+	defer db.Close()
+
+	err = db.CreateCertificateRequest(AppleCSR)
+	if err != nil {
+		t.Fatalf("Couldn't complete Create: %s", err)
+	}
+	err = db.CreateCertificateRequest(BananaCSR)
+	if err != nil {
+		t.Fatalf("Couldn't complete Create: %s", err)
+	}
+	err = db.AddCertificateChainToCertificateRequestByCSR(AppleCSR, AppleCert+IntermediateCert+RootCert)
+	if err != nil {
+		t.Fatalf("Couldn't complete Create: %s", err)
+	}
+	err = db.AddCertificateChainToCertificateRequestByCSR(BananaCSR, BananaCert+IntermediateCert+RootCert)
+	if err != nil {
+		t.Fatalf("Couldn't complete Create: %s", err)
+	}
+	certs, err := db.ListCertificates()
+	if err != nil {
+		t.Fatalf("Couldn't complete List: %s", err)
+	}
+	if len(certs) != 4 {
+		t.Fatalf("Expected 4 Certificates, only got %d", len(certs))
+	}
+	if certs[0].Issuer != 0 || certs[1].Issuer != 1 || certs[2].Issuer != 2 || certs[3].Issuer != 2 {
+		t.Fatalf("Certificate chains were not created correctly")
+	}
+
+	retrievedCSR, err := db.GetCertificateRequestByCSR(AppleCSR)
+	if err != nil {
+		t.Fatalf("Couldn't complete Retrieve: %s", err)
+	}
+	if retrievedCSR.CSR != AppleCSR {
+		t.Fatalf("The CSR from the database doesn't match the CSR that was given")
+	}
+	if retrievedCSR.CertificateID != 3 {
+		t.Fatalf("The certificate chain from the database doesn't match the certificate chain that was given")
+	}
+
+	retrievedCSRWithCert, err := db.GetCertificateRequestAndChainByCSR(AppleCSR)
+	if err != nil {
+		t.Fatalf("Couldn't complete Retrieve: %s", err)
+	}
+	if retrievedCSRWithCert.CertificateChain != AppleCert+"\n"+IntermediateCert+"\n"+RootCert {
+		t.Fatalf("The certificate chain from the database doesn't match the certificate chain that was given")
+	}
+
+	err = db.AddCertificateChainToCertificateRequestByCSR(BananaCSR, BananaCert+IntermediateCert+RootCert)
+	if err != nil {
+		t.Fatalf("Couldn't complete Update: %s", err)
+	}
+	retrievedCSRWithCert, _ = db.GetCertificateRequestAndChainByCSR(BananaCSR)
+	if retrievedCSRWithCert.CertificateChain != BananaCert+"\n"+IntermediateCert+"\n"+RootCert {
+		t.Fatalf("The certificate that was uploaded does not match the certificate that was given.\n Retrieved: %s\nGiven: %s", retrievedCSRWithCert.CertificateChain, BananaCert+IntermediateCert+RootCert)
+	}
+	err = db.RevokeCertificateByCSR(BananaCSR)
+	if err != nil {
+		t.Fatalf("Couldn't complete Update to revoke certificate: %s", err)
+	}
+	retrievedCSR, _ = db.GetCertificateRequestByCSR(BananaCSR)
+	if retrievedCSR.Status != "Revoked" {
+		t.Fatalf("Couldn't revoke certificate")
+	}
+}
+
+func TestUpdateFails(t *testing.T) {
```

Comment:
I think we are lacking more failure scenarios test cases.

Reply:
I added a bit more tests, but it's still not comprehensive because Notary doesn't actually have certificate management capabilities. I am planning on writing a bunch more tests during the notary as a CA task where certificates will be able to be directly managed.